packages/material-ui-styles/src/makeStyles/makeStyles.js

```diff
@@ -241,6 +241,23 @@ export default function makeStyles(stylesOrCreator, options = {}) {
       // eslint-disable-next-line react-hooks/rules-of-hooks
       React.useDebugValue(classes);
     }
+    if (process.env.NODE_ENV !== 'production') {
+      const whitelistedComponents = ['MuiButton'];
```

Comment:
If we want to whitelist then this should be used throughout the codebase. Otherwise this will grow stale.
Why don't we allow it for e.g. TextField?

Reply:
Not all components follow the logic of having the variant be used as `classes[variant]`, the `Typography` and some other components are examples of this. So I wanted to avoid wrong warnings here, by whitelisting which components have custom variants enabled so far. Is there a better place where we can define this?

Reply:
So even though both components have a `variant` prop I couldn't use this API to add another Typography variant? The problem is that one has to be aware of the implementation and technical limitations of the new API. I find this very problematic. Do you have a plan to make it work for Typography as well? Doesn't have to be in this PR (as a MVP) but if we won't be able to reconcile this difference then this API will cause quite a bit of confusion.

At least we should have a descriptive warning if one wants to extend variants in components where this isn't supported.

Reply:
Even now, you can add custom variants to the Typography, because the newly created classes from the variants key are self generated and used in the same manner. The only thing we need to add per component is new type for allowing users to extend the `variant` TS type, as well as relax the `propTypes` for it. Then we can work on the warnings per component if we don't want to introduce breaking changes on how the classes have been generated so far

Tested locally with this example
```
const theme = createMuiTheme({
  variants: {
    MuiTypography: [
      {
        props: { variant: 'custom' },
        styles: { backgroundColor: 'green' }
      }
    ]
  },
});

// usage
<Typography variant="custom">Custom</Typography>
```
![image](https://user-images.githubusercontent.com/4512430/88379544-7553a200-cda3-11ea-922c-9bd9a529f240.png)

Reply:
So you're only adding the whitelist because propTypes and TS would warn about it? But right now the warning is saying that it isn't supported. propTypes and TS are just an interface description. If we agree that this is the correct interface then we don't need this extra allow-list. Otherwise we have to keep track of propTypes/TS in this allow-list which just adds maintenance burden without any benefit as far as I can tell.

Reply:
For implementation of the feature yes, proptypes, TS and the new hook should be added. 

I was trying to keep the warning as simple as possible, that's why it is checking only if the classes[variant] exists. Once we enable it per component we can see if we need to adjust it..

We can complicate more by checking the theme's variants props and try to find a matcher there, or see if the component itself is handling that variant, but I didn't wanted to complicate until it really is necessary.

The `Typography` for example doesn't have classes[variant] for the value `inherit` so this will always throw. That's why I think we will need to anyway adjusts the warning when adding new components.

Let me know if this makes sense.